### PR TITLE
Fix create-testnet-data creating negative supply

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Fix `create-testnet-data` creating negative supply
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
   - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

`create-testnet-data` was creating negative supply which was preventing node from start

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs [hunk 1]
```diff
@@ -19,11 +34,60 @@ hprop_create_testnet_data_minimal =
...
-    -- We test that the command doesn't crash, because otherwise
-    -- execCardanoCLI would fail.
+hprop_create_testnet_data_create_nonegative_supply :: Property
+hprop_create_testnet_data_create_nonegative_supply = do
+  -- FIXME rewrite this as a property test
+  let supplyValues =
+        [ -- (total supply, delegated supply)
+          (2_000_000_000, 1_000_000_000)
+        , (1_000_000_000, 1_000_000_000)
+        , (1_000_000_000, 2_000_000_000)
```
Comment (smelc) [inline cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs]: @carbolymer> when _delegated supply_ is bigger than _total supply_, then `create-testnet-data` should crash. I can do that in a follow-up PR. What do you think?

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs [hunk 2]
```diff
@@ -19,11 +34,60 @@ hprop_create_testnet_data_minimal =
...
+        , "--utxo-keys", "3"
+        , "--drep-keys", "3"
+        , "--out-dir", outputDir
+        ]
+
+      testGenesis@TestGenesis{maxLovelaceSupply, initialFunds} <- H.leftFailM . readJsonFile $ outputDir </> "genesis.json"
+      H.note_ $ show testGenesis
+
+      H.note_ "check that max lovelace supply is positive"
+      H.assertWith maxLovelaceSupply (>= 0)
```
Comment (smelc) [inline cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs]: Can't you check that `maxLovelaceSupply` is the total supply?

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs]: I agree, that would be correct behaviour. But what when they're equal? I was getting negative initial funds for `(1_000_000_000, 1_000_000_000)` case

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs [hunk 3]
```diff
@@ -19,11 +34,60 @@ hprop_create_testnet_data_minimal =
...
     H.noteM_ $ execCardanoCLI
       ["conway",  "genesis", "create-testnet-data"
       , "--testnet-magic", "42"
       , "--out-dir", outputDir
       ]
+    success
 
-    -- We test that the command doesn't crash, because otherwise
-    -- execCardanoCLI would fail.
+hprop_create_testnet_data_create_nonegative_supply :: Property
```
Comment (smelc) [inline cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs:47]:
```suggestion
-- Execute this test with:
-- @cabal test cardano-cli-test --test-options '-p "/create testnet data create nonegative supply/"'@
hprop_create_testnet_data_create_nonegative_supply :: Property
```

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs [hunk 4]
```diff
@@ -619,9 +619,9 @@ updateOutputTemplate
     nonDelegCoin = fromIntegral totalSupply - delegCoin
 
     distribute :: Integer -> Int -> [AddressInEra ShelleyEra] -> [(AddressInEra ShelleyEra, Lovelace)]
-    distribute funds nAddrs addrs = zip addrs (fmap Lovelace (coinPerAddr + remainder:repeat coinPerAddr))
+    distribute funds nAddrs addrs = zip addrs $ Lovelace <$> (coinPerAddr + remainder:repeat coinPerAddr)
       where coinPerAddr, remainder :: Integer
-            (coinPerAddr, remainder) = funds `divMod` fromIntegral nAddrs
+            (coinPerAddr, remainder) = max 0 funds `divMod` fromIntegral nAddrs
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]: This is a stopgap. What we should do is have a check that errors if `funds` is negative.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs]: I've added error throwing when created funds are negative.

File: cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs (hunk 2 above)
Comment (carbolymer) [reply cardano-cli/test/cardano-cli-test/Test/Cli/CreateTestnetData.hs]: done

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs (hunk 4 above)
Comment (carbolymer) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]: Added value check.

**Comment by carbolymer** [comment]
@smelc thanks for pushing the commit with unique worksaces, but I finally reverted it. The reason is that it was circumventing the behaviour of our integration tests, which was saving only failed workspaces by default. After your change all the workspace were persisted which is too much information I think.

Additionally I am not sure about the behaviour `Integration` test when it's executed multiple times as a property test. This needs to be investigated more thoroughly.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs [hunk 5]
```diff
@@ -624,8 +625,10 @@ updateOutputTemplate
     delegCoin = case mDelegatedSupply of Nothing -> 0; Just amountDeleg -> fromIntegral totalSupply - unLovelace amountDeleg
     nonDelegCoin = fromIntegral totalSupply - delegCoin
 
-    distribute :: Integer -> Int -> [AddressInEra ShelleyEra] -> [(AddressInEra ShelleyEra, Lovelace)]
-    distribute funds nAddrs addrs = zip addrs (fmap Lovelace (coinPerAddr + remainder:repeat coinPerAddr))
+    distribute :: Integer -> Int -> [AddressInEra ShelleyEra] -> m [(AddressInEra ShelleyEra, Lovelace)]
+    distribute funds nAddrs addrs
+      | funds < 0 = throwError GenesisCmdNegativeInitialFunds
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]: We can return the exact amount in this error to the user so they can either:
- Decrease the delegated supply
- Increase the total supply

by this amount.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

**Review by smelc — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs (hunk 5 above)
Comment (carbolymer) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs]: Added.

**Review by Jimbo4350 — APPROVED** [review]

**Pull request merged**